doc/dev/adr/0025-rpc-streaming.md

```diff
@@ -0,0 +1,282 @@
+
+[gRPC also allows for either the client or server to cancel an RPC at any time](https://grpc.io/docs/what-is-grpc/core-concepts/#cancelling-an-rpc)
+
+## Decision
+
+### API design, .NET
+
+While RPC streaming shares a lot of similarities to normal RPC, we will define a new communication pattern to handle this scenario with two corresponding base classes: ```StreamingCommandInvoker``` and ```StreamingCommandExecutor```. 
+
+These new base classes will use extended versions of the ```ExtendedRequest``` and ```ExtendedResponse``` classes to include the streaming-specific information about each request and response:
+
+```csharp
+public class StreamingExtendedRequest<TResp> : ExtendedRequest<TResp>
+    where TResp : class
+{
+    /// <summary>
+    /// The index of this request relative to the other requests in this request stream. Starts at 0.
+    /// </summary>
+    public int StreamingRequestIndex { get; set; }
+}
+
+public class StreamingExtendedResponse<TResp> : ExtendedResponse<TResp>
+    where TResp : class
+{
+    /// <summary>
+    /// The index of this response relative to the other responses in this response stream. Starts at 0.
+    /// </summary>
+    public int StreamingResponseIndex { get; set; }
+}
+```
+
+#### Invoker side
+
+The new ```StreamingCommandInvoker``` will largely look like the existing ```CommandInvoker```, but will instead have an API for ```InvokeCommandWithStreaming```.
+
+This new method will take the same parameters as ```InvokeCommand``` but will accept a stream of requests and return a stream of command responses. 
+
+```csharp
+public abstract class StreamingCommandInvoker<TReq, TResp>
+    where TReq : class
+    where TResp : class
+{
+    // Many requests, many responses.
+    public IAsyncEnumerable<StreamingExtendedResponse<TResp>> InvokeStreamingCommandAsync(IAsyncEnumerable<TReq> requests, ...) {...}
+}
+```
+
+#### Executor side
+
+The new ```StreamingCommandExecutor``` will largely look like the existing ```CommandExecutor```, but the callback to notify users that a command was received will include a stream of requests and return a stream of responses.
+
+```csharp
+public abstract class StreamingCommandExecutor<TReq, TResp> : IAsyncDisposable
+    where TReq : class
+    where TResp : class
+{
+    /// <summary>
+    /// A streaming command was invoked
+    /// </summary>
+    /// <remarks>
+    /// The callback provides the stream of requests and requires the user to return one to many responses.
+    /// </remarks>
+    public required Func<IAsyncEnumerable<StreamingExtendedRequest<TReq>>, CancellationToken, IAsyncEnumerable<StreamingExtendedResponse<TResp>>> OnStreamingCommandReceived { get; set; }
+}
+
+```
+
+With this design, commands that use streaming are defined at codegen time. Codegen layer changes will be defined in a separate ADR, though.
```

Comment:
The ProtocolCompiler will certainly need substantial additions to handle this design, but I think the DTDL additions could be nearly trivial.  The structural similarities are so strong that I think the current DTDL Command would be sufficient.  We'd just need to add a simple co-type, such as Streaming, to indicate that the codegen target is the SDK's streaming RPC API rather than the normal RPC API.